manager/orchestrator/taskreaper/task_reaper.go

```diff
@@ -269,11 +269,9 @@ func (tr *TaskReaper) tick() {
 			for _, t := range historicTasks {
 				// Skip tasks which are desired to be running but the current state
 				// is less than or equal to running.
-				// This check is important to ignore tasks which are running or need to be running,
-				// but to delete tasks which are either past running,
-				// or have not reached running but need to be shutdown (because of a service update, for example).
-				if t.DesiredState == api.TaskStateRunning && t.Status.State <= api.TaskStateRunning {
-					// Don't delete running tasks
+				// Ignore tasks which are running (including tasks which are desired to be shutdown),
+				// or which are desired to be running (desired state running).
+				if t.Status.State == api.TaskStateRunning || t.DesiredState <= api.TaskStateRunning {
```

Comment:
What about e.g. (state=starting, desired=shutdown)? We shouldn't reap that.

As I understand it, the reaper is only allowed to remove a task when:
- `state >= completed` (resources freed), OR
- `state < assigned` AND `desired >= completed` (resources will never be assigned on a node)

Reply:
fair point. I think a task can still start even though it hasn't yet and even though it has been marked for shutdown, so it shouldn't be reaped in that case.

Reply:
For readability, perhaps these should be split over two if-statements; it allows describing/documenting per condition, which may help understanding the logic)

```go
if t.Status.State == api.TaskStateRunning {
	runningTasks++
	continue
}
if t.DesiredState <= api.TaskStateRunning {
	runningTasks++
	continue
}
```

or a `switch`, whatever looks best 👍 

```go
switch {
case t.Status.State == api.TaskStateRunning:
	runningTasks++
	continue

case  t.DesiredState <= api.TaskStateRunning:
	runningTasks++
	continue
}
```